crates/p4-analyzer/src/driver.rs

```diff
@@ -0,0 +1,233 @@
+		let match_func = || {
+			match input_source.clone() {
+				DriverType::Console => Message::read(&mut stdin().lock()),
+				#[cfg(test)] 
+				DriverType::Buffer(mut buffer) => block_on(buffer.message_read()),
+		}};
+
+		while let Ok(Some(message)) = match_func() {
+			if sender.send_blocking(message).is_err() {
+				break;
+			}
+		}
+	}
+
+	fn receiver_task(receiver: Receiver<Message>, output_source: DriverType) {
+		let message_send = |message: Message| {
+			match output_source.clone() {
+				DriverType::Console => message.write(&mut stdout()),
+				#[cfg(test)] 
+				DriverType::Buffer(mut buffer) => block_on(buffer.message_write(message)),
+		}};
+
+		while let Ok(message) = receiver.recv_blocking() {
+			if message_send(message).is_err() {
+				break;
+			}
+		}
+	}
+
+	/// Starts executing the [`ConsoleDriver`] instance.
+	pub async fn start(&self, cancel_token: Arc<CancellationToken>) -> Result<(), OperationCanceled> {
+		let (sender, _) = self.in_channel.clone();
+		let (_, receiver) = self.out_channel.clone();
+
+		let source_type = self.source_type.clone();
+		let source_type2 = self.source_type.clone();
+		let _sender_task = std::thread::spawn(move || Self::sender_task(sender, source_type));
+		let receiver_task = std::thread::spawn(move || Self::receiver_task(receiver, source_type2));
+
+		let (sender, _) = self.in_channel.clone();
+		let (_, receiver) = self.out_channel.clone();
+
+		// Joining to the `receiver_task` will block the current thread. Since this will likely be the main thread it will
+		// prevent the async Futures from being driven forward. `spawn_blocking` allows this blocking code to be taken into
+		// another thread and returns a `Future` that we can then await.
+		task::spawn_blocking(move || {
+			cancel_token.run(
+				|| {
+					sender.close();
+					receiver.close();
+				},
+				|| {
+					receiver_task.join().unwrap();
+
+					match cancel_token.is_canceled() {
+						true => Err(OperationCanceled),
+						_ => Ok(()),
+					}
+				},
+			)
+		})
+		.await
+		.unwrap()
+	}
+}
+
+/// All code below is for DriveType::Buffer(BufferStruct)
+/// This type is used for testing but due to needing of talking between threads it has a heavy implementation
+
+#[cfg(test)]
+use queues::*;
+#[cfg(test)]
+use std::{
+		io::{self},
+		sync::RwLock,
+		time::Duration
+	};
+
+#[cfg(test)]
+pub fn buffer_driver(buffer: BufferStruct) -> Driver {
+	Driver {
+		in_channel: async_channel::unbounded::<Message>(),
+		out_channel: async_channel::unbounded::<Message>(),
+		source_type: DriverType::Buffer(buffer),
+	}
+}
+#[cfg(test)]
+struct BufferStructData {
+	input_queue: Queue<Message>,	// stores the messages to be send
+	output_buffer: Vec<u8>,			// stores the recieved messages(write! doesn't seperate each message)
+	output_count: usize,			// Tell you how many messages recieved
+	read_queue_count: usize,		// Tells the thread to send a message if var isn't 0
+}
```

Comment:
I'm not sure about `read_queue_count`'s comment, isn't it instructing the thread to _read_ a message from the queue?
```suggestion
#[cfg(test)]
struct BufferStructData {
	input_queue: Queue<Message>,	// stores the messages to be sent
	output_buffer: Vec<u8>,			// stores the received messages (write! doesn't seperate messages)
	output_count: usize,			// the number of messages received so far
	read_queue_count: usize,		// tells the thread to send a message if var isn't 0
}
```